vortex-array/src/array/primitive/compute/fill_null.rs

```diff
@@ -0,0 +1,132 @@
+use vortex_buffer::Buffer;
+use vortex_dtype::{match_each_native_ptype, Nullability};
+use vortex_error::{vortex_err, VortexExpect, VortexResult};
+
+use crate::array::primitive::PrimitiveArray;
+use crate::array::{ConstantArray, PrimitiveEncoding};
+use crate::compute::FillNullFn;
+use crate::validity::{ArrayValidity, Validity};
+use crate::variants::PrimitiveArrayTrait;
+use crate::{ArrayDType, ArrayData, ArrayLen, IntoArrayData, ToArrayData};
+
+impl FillNullFn<PrimitiveArray> for PrimitiveEncoding {
+    fn fill_null(
+        &self,
+        array: &PrimitiveArray,
+        fill_value: vortex_scalar::Scalar,
+    ) -> VortexResult<ArrayData> {
+        let result_validity = match fill_value.dtype().nullability() {
+            Nullability::NonNullable => Validity::NonNullable,
+            Nullability::Nullable => Validity::AllValid,
+        };
+
+        if array.dtype().nullability() == Nullability::NonNullable
+            && fill_value.dtype().nullability() == Nullability::NonNullable
+        {
+            return Ok(array.to_array());
+        }
+
+        let array_validity = array.logical_validity();
+        if array_validity.all_valid() {
+            return Ok(PrimitiveArray::from_byte_buffer(
+                array.byte_buffer().clone(),
+                array.ptype(),
+                result_validity,
+            )
+            .into_array());
+        }
+
+        if array_validity.all_invalid() {
+            return Ok(ConstantArray::new(fill_value, array.len()).into_array());
+        }
+
+        let nulls = array_validity
+            .to_null_buffer()?
+            .ok_or_else(|| vortex_err!("Failed to convert array validity to null buffer"))?;
+
+        // TODO(danking): when we take PrimitiveArray by value, we should mutate in-place
+        match_each_native_ptype!(array.ptype(), |$T| {
+            let as_slice = array.as_slice::<$T>();
+            let fill_value = fill_value
+                .as_primitive()
+                .typed_value::<$T>()
+                .vortex_expect("top-level fill_null ensure non-null fill value");
+            let filled = Buffer::from_iter(
```

Comment:
I think it would be faster to just iterate validity and set fileds in the new buffer?

Reply:
Yep. Done.